flash/audio/speech_recognition/data.py

```diff
@@ -0,0 +1,291 @@
+
+
+class SpeechRecognitionJSONDataSource(SpeechRecognitionDataSource):
+
+    def __init__(self, backbone: str):
+        super().__init__(backbone, filetype='json')
+
+
+class SpeechRecognitionFilesSource(DataSource):
+
+    def __init__(self, backbone: str, filetype: Optional[str] = None):
+        super().__init__()
+        self.filetype = filetype
+        self.backbone = backbone
+
+    def load_data(
+        self,
+        files: Tuple[str, Union[str, List[str]], Union[str, List[str]]],
+        dataset: Optional[Any] = None,
+    ) -> Union[Sequence[Mapping[str, Any]]]:
+        if isinstance(files, str):
+            files = [files]
+        return [dict(file=file) for file in files]
+
+
+class SpeechRecognitionPreprocess(Preprocess):
+
+    def __init__(
+        self,
+        train_transform: Optional[Dict[str, Callable]] = None,
+        val_transform: Optional[Dict[str, Callable]] = None,
+        test_transform: Optional[Dict[str, Callable]] = None,
+        predict_transform: Optional[Dict[str, Callable]] = None,
+        backbone: str = "facebook/wav2vec2-base-960h",
+    ):
+        self.backbone = backbone
+
+        super().__init__(
+            train_transform=train_transform,
+            val_transform=val_transform,
+            test_transform=test_transform,
+            predict_transform=predict_transform,
+            data_sources={
+                DefaultDataSources.CSV: SpeechRecognitionCSVDataSource(self.backbone),
+                DefaultDataSources.JSON: SpeechRecognitionJSONDataSource(self.backbone),
+                "timit": TimitDataSource(self.backbone),
+                DefaultDataSources.FILES: SpeechRecognitionFilesSource(self.backbone)
+            },
+            default_data_source=DefaultDataSources.FILES,
+            deserializer=SpeechRecognitionDeserializer(backbone),
+        )
+        self.processor = Wav2Vec2Processor.from_pretrained(backbone)
+        self.collator = DataCollatorCTCWithPadding(processor=self.processor, padding=True)
+
+    def get_state_dict(self) -> Dict[str, Any]:
+        return {
+            **self.transforms,
+            "backbone": self.backbone,
+        }
+
+    @classmethod
+    def load_state_dict(cls, state_dict: Dict[str, Any], strict: bool):
+        return cls(**state_dict)
+
+    def _prepare_dataset(self, batch: Any) -> Any:
+        # check that all files have the correct sampling rate
+        assert (
+            len(set(batch["sampling_rate"])) == 1
+        ), f"Make sure all inputs have the same sampling rate of {self.processor.feature_extractor.sampling_rate}."
+
+        batch["input_values"] = self.processor(batch["speech"], sampling_rate=batch["sampling_rate"][0]).input_values
+
+        if not self.predicting:
+            with self.processor.as_target_processor():
+                batch["labels"] = self.processor(batch["target_text"]).input_ids
+        return batch
+
+    def _speech_file_to_array_fn(self, batch: Any) -> Any:
```

Comment:
By putting the read file logic in the collating logic, we save processing the entire dataset upfront, and only loading what's necessary. It also means we cover multi-processing the batches via the data-loader which is nice :)